lib/Word_Lib/Boolean_Inequalities.thy

```diff
@@ -0,0 +1,141 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+theory Boolean_Inequalities
+  imports Word_EqI
+begin
+
+section \<open>All inequalities between binary Boolean operations on @{typ "'a word"}\<close>
+
+text \<open>
+  Enumerates all binary functions resulting from Boolean operations on @{typ "'a word"}
+  and derives all inequalities of the form @{term "f x y \<le> g x y"} between them.
+  We leave out the trivial @{term "0 \<le> g x y"}, @{term "f x y \<le> -1"}, and
+  @{term "f x y \<le> f x y"}, because these are already readily available to the simplifier and
+  other methods.
+
+  This leaves 36 inequalities. Some of these are subsumed by each other, but we generate
+  the full list to avoid too much manual processing.
+
+  All inequalities produced here are in simp normal form.\<close>
```

Comment:
This is a really cool idea. How did you come up with it?

Reply:
There was a table of all boolean operators in the book and a description that you can read off `<=` from that. The description was implication (without saying the name), and from that it wasn't a big step to generating them all instead of trying to write them down.

I've been thinking about generating all 3-argument boolean operators as well (same principle), but it's probably overkill for now.